src/index.py

```diff
@@ -79,3 +96,269 @@ def contents_from_page(page: str) -> str:
+
+
+def _iter_index_lines(lines: typing.Iterable[str], section: _IndexSection) -> typing.Iterator[str]:
+    """Remove the contents section lines from the index contents.
+
+    Args:
+        lines: The lines of the index file.
+        section: The part of the index file to return lines for.
+
+    Yields:
+        All lines except the lines of the contents section.
+    """
+    contents_encountered = False
+    drop_lines = section == _IndexSection.CONTENTS
+    for line in lines:
+        if not contents_encountered and line.lower() == CONTENTS_HEADER:
+            contents_encountered = True
+            drop_lines = section == _IndexSection.EX_CONTENTS
+        elif line.startswith(CONTENTS_END_LINE_PREFIX):
+            drop_lines = section == _IndexSection.CONTENTS
+
+        if not drop_lines:
+            yield line
+
+
+def get_content_for_server(index_file: IndexFile) -> str:
+    """Get the contents from the index file that should be passed to the server.
+
+    Args:
+        index_file: Information about the local index file.
+
+    Returns:
+        The contents of the index file that should be stored on the server.
+    """
+    if index_file.content is None:
+        return ""
+
+    return "\n".join(
+        _iter_index_lines(index_file.content.splitlines(), section=_IndexSection.EX_CONTENTS)
+    )
+
+
+def _get_contents_parsed_items(index_file: IndexFile) -> typing.Iterator[_ParsedListItem]:
+    """Get the items from the contents list of the index file.
+
+    Args:
+        index_file: The index file to read the contents from.
+
+    Yields:
+        All the items on the contents list in the index file.
+    """
+    if index_file.content is None:
+        return
+
+    # Get the lines of the contents section
+    contents_lines = _iter_index_lines(
+        index_file.content.splitlines(), section=_IndexSection.CONTENTS
+    )
+    # Skip header
+    next(contents_lines, None)
+    yield from (
+        _parse_item_from_line(line=line, rank=rank)
+        for line, rank in zip(filter(None, contents_lines), itertools.count())
+    )
+
+
+def _check_contents_item(
+    item: _ParsedListItem, whitespace_expectation: int, aggregate_dir: Path, docs_path: Path
+) -> None:
+    """Check item is valid. All the items should be exactly within a directory.
+
+    Args:
+        item: The parsed item to check.
+        whitespace_expectation: The expected number of whitespace characters for items.
+        aggregate_dir: The relative directory that all items must be within.
+        docs_path: The base directory of all items.
+
+    Raises:
+        InputError:
+            - An item has more whitespace than a previous item and it is not following a directory.
+            - A nested item is not immediately within the path of its parent.
+            - An item isn't a file nor directory.
+    """
+    # Check that the whitespace count matches the expectation
+    if item.whitespace_count > whitespace_expectation:
```

Comment:
Out of curiosity, why is this not a non equality operator, as in `if item.whitespace_count != whitespace_expectation:`? If the condition is intended, would `max_expected_whitespaces` help describe it clearer?

Reply:
Yeh that is fair feedback, I'll change it to max whitespace to make it clearer

Reply:
Should be fixed now